presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -401,6 +417,27 @@ private void processNewSources()
                 }
             }
 
+            for (int index = operators.size() - 1; index >= 0; index--) {
+                if (operators.get(index).isFinished()) {
+                    // close and remove this operator and all source operators
+                    List<Operator> finishedOperators = this.operators.subList(0, index + 1);
```

Comment:
`finishedOperators` -- they are not necessarily finished (except for the last one). `obsoleteOperators`?